bme280/bme280.go

```diff
@@ -0,0 +1,121 @@
+package bme280
+
+import (
+	"machine"
+	"time"
+)
+
+// calibrationCoefficients reads at startup and stores the calibration coefficients
+type calibrationCoefficients struct {
+	t1 uint16
+	t2 int16
+	t3 int16
+	p1 uint16
+	p2 int16
+	p3 int16
+	p4 int16
+	p5 int16
+	p6 int16
+	p7 int16
+	p8 int16
+	p9 int16
+}
+
+// Device wraps an I2C connection to a BME280 device.
+type Device struct {
+	bus                     machine.I2C
+	Address                 uint16
+	calibrationCoefficients calibrationCoefficients
+}
+
+// New creates a new BME280 connection. The I2C bus must already be
+// configured.
+//
+// This function only creates the Device object, it does not touch the device.
+func New(bus machine.I2C) Device {
+	return Device{
+		bus:     bus,
+		Address: Address,
+	}
+}
+
+// Configure sets up the device for communication and
+// read the calibration coefficientes.
+func (d *Device) Configure() {
+
+	data := make([]byte, 24)
+	err := d.bus.ReadRegister(uint8(d.Address), REG_CALIBRATIION, data)
+	if err != nil {
+		return
+	}
+	d.calibrationCoefficients.t1 = readUint(data[0], data[1])
+	d.calibrationCoefficients.t2 = readInt(data[2], data[3])
+	d.calibrationCoefficients.t3 = readInt(data[4], data[5])
+	d.calibrationCoefficients.p1 = readUint(data[6], data[7])
+	d.calibrationCoefficients.p2 = readInt(data[8], data[9])
+	d.calibrationCoefficients.p3 = readInt(data[10], data[11])
+	d.calibrationCoefficients.p4 = readInt(data[12], data[13])
+	d.calibrationCoefficients.p5 = readInt(data[14], data[15])
+	d.calibrationCoefficients.p6 = readInt(data[16], data[17])
+	d.calibrationCoefficients.p7 = readInt(data[18], data[19])
+	d.calibrationCoefficients.p8 = readInt(data[20], data[21])
+	d.calibrationCoefficients.p9 = readInt(data[22], data[23])
+
+	d.bus.WriteRegister(uint8(d.Address), CTRL_MEAS_ADDR, []byte{0x25})
+	d.bus.WriteRegister(uint8(d.Address), CTRL_HUMIDITY_ADDR, []byte{0x01})
+	d.bus.WriteRegister(uint8(d.Address), CTRL_CONFIG, []byte{0xA8})
+
+}
+
+// Connected returns whether a BME280 has been found.
+// It does a "who am I" request and checks the response.
+func (d *Device) Connected() bool {
+	data := []byte{0}
+	d.bus.ReadRegister(uint8(d.Address), WHO_AM_I, data)
+	time.Sleep(1 * time.Second)
+	return data[0] == CHIP_ID
+}
+
+// Temperature returns the temperature in celsius milli degrees (ºC/1000)
+func (d *Device) Temperature() (temperature float32, err error) {
```

Comment:
I suggest changing the signature to this:

```go
func (d *Device) Temperature() (temperature int32, err error)
```

This lets you match the other `Temperature()` methods on other sensors. There is not a formal interface, but this is the custom we have been trying to follow.

Please let me know if this makes sense.

Reply:
At the moment we have both `Temperature() (temperature int32, err error)` and `ReadTemperature() (int32, error)` in the code base (and `ReadTemperature() (temp int8)` but that should be updated). My preference goes to `ReadTemperature() (int32, error)` because the `Read` prefix makes it clear the operation does I/O (and isn't just a getter). Also see https://github.com/tinygo-org/drivers/wiki/Driver-design---API#simple-sensors

Reply:
I think we should pick a single interface and stick with that. We can bikeshed about which it is, but it should be the same across all drivers (where reasonable).

Reply:
OK, let's go with `ReadTemperature() (int32, error)` across all drivers. Change made.